File: mobile/backend/app/repositories/operation_repository.py
"""
Operation repository for Firestore operations.
"""

from typing import List, Optional, Dict, Any

from app.repositories.base_repository import BaseRepository
from app.core.enums import OperationType, OperationStatus


class OperationRepository(BaseRepository):
    """Repository for Operation documents."""

    def __init__(self):
        super().__init__("operations")

    async def get_by_employee(self, employee_id: str) -> List[Dict[str, Any]]:
        """Get all operations assigned to an employee."""
        return await self.query(filters=[("employee_id", "==", employee_id)])

    async def get_by_order(self, order_id: str) -> List[Dict[str, Any]]:
        """Get all operations for a specific order."""
        return await self.query(filters=[("order_id", "==", order_id)])

    async def get_by_status(self, status: OperationStatus) -> List[Dict[str, Any]]:
        """Get operations by status."""
        return await self.query(filters=[("status", "==", status.value)])

    async def get_pending_for_employee(self, employee_id: str) -> List[Dict[str, Any]]:
        """Get pending operations for a specific employee."""
        return await self.query(filters=[
            ("employee_id", "==", employee_id),
            ("status", "==", OperationStatus.PENDING.value),
        ])

    async def get_validated(self) -> List[Dict[str, Any]]:
        """Get all validated operations (validated_at is not null)."""
        all_ops = await self.get_all()
        return [op for op in all_ops if op.get("validated_at")]

    async def get_filtered(
        self,
        operation_type: Optional[OperationType] = None,
        employee_id: Optional[str] = None,
        status: Optional[OperationStatus] = None,
    ) -> List[Dict[str, Any]]:
        """Get operations with optional filters."""
        filters = []
        if operation_type:
            filters.append(("type", "==", operation_type.value))
        if employee_id:
            filters.append(("employee_id", "==", employee_id))
        if status:
            filters.append(("status", "==", status.value))
        return await self.query(filters=filters if filters else None)

File: mobile/frontend/lib/data/repositories/operation_repository.dart
import 'package:frontend/data/models/operation.dart';
import 'package:frontend/data/enums.dart';
import 'base_repository.dart';

class OperationRepository extends BaseRepository<Operation> {
  OperationRepository() : super('operations');

  @override
  Operation fromMap(Map<String, dynamic> map) => Operation.fromMap(map);

  @override
  Map<String, dynamic> toMap(Operation entity) => entity.toMap();

  // ── Domain queries ────────────────────────────────────────

  /// Get operations by type.
  Future<List<Operation>> getByType(OperationType type) {
    return query(
      where: 'type = ?',
      whereArgs: [type.value],
      orderBy: 'created_at DESC',
    );
  }

  /// Get operations by status.
  Future<List<Operation>> getByStatus(OperationStatus status) {
    return query(
      where: 'status = ?',
      whereArgs: [status.value],
      orderBy: 'created_at DESC',
    );
  }

  /// Get operations assigned to an employee.
  Future<List<Operation>> getByEmployee(String employeeId) {
    return query(
      where: 'employee_id = ?',
      whereArgs: [employeeId],
      orderBy: 'created_at DESC',
    );
  }

  /// Get pending operations for an employee.
  Future<List<Operation>> getPendingForEmployee(String employeeId) {
    return query(
      where: "employee_id = ? AND status IN ('pending', 'in_progress')",
      whereArgs: [employeeId],
      orderBy: 'created_at ASC',
    );
  }

  /// Get operations linked to an order.
  Future<List<Operation>> getByOrder(String orderId) {
    return query(
      where: 'order_id = ?',
      whereArgs: [orderId],
      orderBy: 'created_at ASC',
    );
  }

  /// Get all operations sorted by date.
  Future<List<Operation>> getAllSorted() {
    return getAll(orderBy: 'created_at DESC');
  }

  /// Get pending operations (all employees).
  Future<List<Operation>> getAllPending() {
    return query(
      where: "status IN ('pending', 'in_progress')",
      orderBy: 'created_at ASC',
    );
  }

  /// Start an operation.
  Future<void> start(String operationId) {
    return update(operationId, {
      'status': OperationStatus.inProgress.value,
      'started_at': DateTime.now().toIso8601String(),
    }).then((_) {});
  }

  /// Complete an operation.
  Future<void> complete(String operationId, {String? validatorId}) {
    final fields = <String, dynamic>{
      'status': OperationStatus.completed.value,
      'completed_at': DateTime.now().toIso8601String(),
    };
    if (validatorId != null) {
      fields['validator_id'] = validatorId;
      fields['validated_at'] = DateTime.now().toIso8601String();
    }
    return update(operationId, fields).then((_) {});
  }

  /// Fail an operation.
  Future<void> fail(String operationId) {
    return update(operationId, {
      'status': OperationStatus.failed.value,
      'completed_at': DateTime.now().toIso8601String(),
    }).then((_) {});
  }

  /// Get operations by chariot.
  Future<List<Operation>> getByChariot(String chariotId) {
    return query(
      where: 'chariot_id = ?',
      whereArgs: [chariotId],
      orderBy: 'created_at DESC',
    );
  }

  /// Count operations by status.
  Future<Map<OperationStatus, int>> countByStatus() async {
    final db = await database;
    final results = await db.rawQuery(
      'SELECT status, COUNT(*) as cnt FROM operations WHERE is_deleted = 0 GROUP BY status',
    );
    final map = <OperationStatus, int>{};
    for (final row in results) {
      map[OperationStatus.fromString(row['status'] as String)] =
          row['cnt'] as int;
    }
    return map;
  }
}
